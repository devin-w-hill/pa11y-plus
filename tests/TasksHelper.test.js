let tasksMod = require('../TasksHelper.js');
let fs = require('graceful-fs');

describe('TasksHelper module tests', () => {
    jest.setTimeout(30000);

    let mockPa11yResult = {
        "documentTitle": "Help - Help - Govdemo",
        "pageUrl": "https://govdemo.staging.tmorders.dev/portal/help/index",
        "issues": [
            {
                "code": "WCAG2AA.Principle2.Guideline2_4.2_4_1.H64.1",
                "type": "error",
                "typeCode": 1,
                "message": "Iframe element requires a non-empty title attribute that identifies the frame.",
                "context": "<iframe style=\"position: absolute; top: 0; left: 0; width: 100%; height: 100%;\" src=\"https://player.vimeo.com/video/398847860?color=ffffff&amp;byline=0&amp;portrait=0\" frameborder=\"0\" allowfullscreen=\"\"></iframe>",
                "selector": "#orderModal > div > div > div > iframe",
                "runner": "htmlcs",
                "runnerExtras": {}
            }
        ]
    }

    beforeEach(() => {
        jest.resetModules();
        jest.resetAllMocks();
    });
    afterEach(() => {
        jest.resetModules();
        jest.resetAllMocks();
        fs = require('graceful-fs');
        tasksMod = require('../TasksHelper.js');
    });

    test('getTasks with invalid file path', () => {
        let filePath = './test/path/to/file.json';
        expect(() => {
            let tasks = tasksMod.getTasks(filePath);
        }).toThrowError('There was no directory or file found at ' + filePath);
    });

    test('getTasks -> directory path -> 0 tasks', () => {
        let filePath = './test/path/';
        let mockFn = jest.fn();
        mockFn.mockReturnValue(true);
        let mockLstat = jest.fn();
        let isDirectory = jest.fn();
        isDirectory.mockReturnValue(true);
        mockLstat.mockReturnValue({
            isDirectory: isDirectory
        });
        fs.existsSync = mockFn;
        fs.lstatSync = mockLstat;
        let mockGetFilesRecursively = jest.fn();
        mockGetFilesRecursively.mockReturnValue([]); // 0 task file paths.
        tasksMod.getFilesRecursively = mockGetFilesRecursively;

        let tasks = tasksMod.getTasks(filePath);
        expect(tasks.length).toBe(0);
    });

    test('getTasks -> directory path -> 2 files with 4 tasks', () => {
        let filePath = './test/path/to/file.json';
        let mockFn = jest.fn();
        mockFn.mockReturnValue(true);
        let mockLstat = jest.fn();
        let isDirectory = jest.fn();
        isDirectory.mockReturnValue(true);
        mockLstat.mockReturnValue({
            isDirectory: isDirectory
        });
        fs.existsSync = mockFn;
        fs.lstatSync = mockLstat;
        let mockGetFilesRecursively = jest.fn();
        mockGetFilesRecursively.mockReturnValueOnce([
            'test/file/path1.json',
            'test/file/path2.json'
        ]); // 0 task file paths.
        tasksMod.getFilesRecursively = mockGetFilesRecursively;
        let mockTraverseFiles = jest.fn();
        mockTraverseFiles.mockReturnValue(['mockTaskObj', 'mockTaskObj2']);
        tasksMod.traverseFiles = mockTraverseFiles;

        let tasks = tasksMod.getTasks(filePath);
        expect(tasks.length).toBe(4); // 4 because there are two file paths, which calls traverseFiles twice.
    });

    test('getTasks -> single file path -> with 1 task object', () => {
        let filePath = './test/path/to/file.json';
        let mockFn = jest.fn();
        mockFn.mockReturnValueOnce(false)
              .mockReturnValueOnce(true);
        fs.existsSync = mockFn;
        let mockReadFileSync = jest.fn();
        mockReadFileSync.mockReturnValueOnce(JSON.stringify([{name: "mockTask"}, {filePath :'notNull'}]));
        fs.readFileSync = mockReadFileSync;

        let tasks = tasksMod.getTasks(filePath);
        expect(tasks.length).toBe(2);
    });

    test('getTasks -> single file path -> with 2 file paths and 4 tasks', () => {
        let filePath = './test/path/to/file.json';
        let mockFn = jest.fn();
        mockFn.mockReturnValueOnce(false)
            .mockReturnValueOnce(true);
        fs.existsSync = mockFn;
        let mockReadFileSync = jest.fn();
        mockReadFileSync.mockReturnValueOnce(JSON.stringify([
            'test/file/path1.json',
            'test/file/path2.json'
        ]));
        fs.readFileSync = mockReadFileSync;
        let mockTraverseFiles = jest.fn();
        mockTraverseFiles.mockReturnValue(['mockTaskObj', 'mockTaskObj2']);
        tasksMod.traverseFiles = mockTraverseFiles;

        let tasks = tasksMod.getTasks(filePath);
        expect(tasks.length).toBe(4); // 4 because there are two file paths, which calls traverseFiles twice.
    });

    test('getTasks -> single file path -> with invalid task file', () => {
        let filePath = './test/path/to/file.json';
        let mockFn = jest.fn();
        mockFn.mockReturnValueOnce(false)
            .mockReturnValueOnce(true);
        fs.existsSync = mockFn;
        let mockReadFileSync = jest.fn();
        mockReadFileSync.mockReturnValueOnce(JSON.stringify([0, 1]));
        fs.readFileSync = mockReadFileSync;

        expect(() => {
            let tasks = tasksMod.getTasks(filePath);
        }).toThrowError('The provided file is invalid. It does not contain a task(s) in valid JSON or a list of filepaths. Filepath: ' + filePath);
    });

    test('getFilesRecursively with single directory depth', () => {
        let mockDirectory = 'test/directory/';
        let mockReaddirSync = jest.fn();
        mockReaddirSync.mockReturnValueOnce([
            'path1.json',
            'path2.json'
        ]);
        fs.readdirSync = mockReaddirSync;
        let mockStatSync = jest.fn();
        let mockIsDirectory = jest.fn();
        mockIsDirectory.mockReturnValueOnce(false);
        mockStatSync.mockReturnValue({
            isDirectory: mockIsDirectory
        });
        fs.statSync = mockStatSync;

        let files = tasksMod.getFilesRecursively(mockDirectory, []);
        expect(files.length).toBe(2);
    });

    test('getFilesRecursively with multiple sub-directories', () => {
        let mockDirectory = 'test/directory/';
        let mockReaddirSync = jest.fn();
        mockReaddirSync.mockReturnValueOnce([
            'path1.json',
            'path2.json',
            'sub/directory/',
            'sub/directory/'
        ]);
        mockReaddirSync.mockReturnValueOnce([
            'path3.json',
            'path4.json'
        ]);
        mockReaddirSync.mockReturnValueOnce([
            'path5.json',
            'path6.json'
        ])
        fs.readdirSync = mockReaddirSync;
        let mockStatSync = jest.fn();
        let mockIsDirectory = jest.fn();
        mockIsDirectory.mockReturnValueOnce(false);
        mockIsDirectory.mockReturnValueOnce(false);
        mockIsDirectory.mockReturnValueOnce(true);
        mockIsDirectory.mockReturnValueOnce(false);
        mockIsDirectory.mockReturnValueOnce(false);
        mockIsDirectory.mockReturnValueOnce(true);
        mockIsDirectory.mockReturnValueOnce(false);
        mockIsDirectory.mockReturnValueOnce(false);
        mockStatSync.mockReturnValue({
            isDirectory: mockIsDirectory
        });
        fs.statSync = mockStatSync;

        let files = tasksMod.getFilesRecursively(mockDirectory, []);
        expect(files.length).toBe(6);
    });

    test('traverseFiles -> directory path -> 0 tasks', () => {
        let filePath = './test/path/';
        let mockExistsSync = jest.fn();
        mockExistsSync.mockReturnValueOnce(true);
        let mockLstat = jest.fn();
        let isDirectory = jest.fn();
        isDirectory.mockReturnValueOnce(true);
        mockLstat.mockReturnValueOnce({
            isDirectory: isDirectory
        });
        fs.existsSync = mockExistsSync;
        fs.lstatSync = mockLstat;
        let mockGetFilesRecursively = jest.fn();
        mockGetFilesRecursively.mockReturnValueOnce([]); // 0 task file paths.
        tasksMod.getFilesRecursively = mockGetFilesRecursively;

        let tasks = tasksMod.traverseFiles(filePath);
        expect(tasks.length).toBe(0);
    });

    test('traverseFiles -> directory path -> 2 tasks', () => {
        let filePath = './test/path/';
        let mockExistsSync = jest.fn();
        mockExistsSync.mockReturnValueOnce(true);
        mockExistsSync.mockReturnValueOnce(true);
        mockExistsSync.mockReturnValueOnce(true);
        let mockLstat = jest.fn();
        let isDirectory = jest.fn();
        isDirectory.mockReturnValue(true);
        mockLstat.mockReturnValue({
            isDirectory: isDirectory
        });
        fs.existsSync = mockExistsSync;
        fs.lstatSync = mockLstat;
        let mockGetFilesRecursively = jest.fn();
        mockGetFilesRecursively.mockReturnValueOnce([
            'file1.json',
            'file2.json'
        ]);
        tasksMod.getFilesRecursively = mockGetFilesRecursively;
        let mockReadFileSync = jest.fn();
        mockReadFileSync.mockReturnValue(JSON.stringify(['mockTaskObject', {filePath: 'notNull'}]));
        fs.readFileSync = mockReadFileSync;

        let tasks = tasksMod.traverseFiles(filePath);
        expect(tasks.length).toBe(4);
    });

    test('traverseFiles -> single file path -> 2 tasks', () => {
        let filePath = './test/path/to/file.json';
        let mockExistsSync = jest.fn();
        mockExistsSync.mockReturnValueOnce(false);
        mockExistsSync.mockReturnValueOnce(true);
        fs.existsSync = mockExistsSync;
        let mockReadFileSync = jest.fn();
        mockReadFileSync.mockReturnValue(JSON.stringify(['mockTaskObject', {filePath: 'notNull' }]));
        fs.readFileSync = mockReadFileSync;

        let tasks = tasksMod.traverseFiles(filePath);
        expect(tasks.length).toBe(2);
    });

    test('traverseFiles -> invalid file path', () => {
        let filePath = './test/path/to/file.json';
        let mockExistsSync = jest.fn();
        mockExistsSync.mockReturnValueOnce(false);
        mockExistsSync.mockReturnValueOnce(false);
        fs.existsSync = mockExistsSync;


        expect(() => {
            let tasks = tasksMod.traverseFiles(filePath);
        }).toThrowError('There was no directory or file found at ' + filePath);
    });

    test('traverseFiles -> invalid file path within a directory', () => {
        let filePath = './test/path/';
        let mockExistsSync = jest.fn();
        mockExistsSync.mockReturnValueOnce(true);
        mockExistsSync.mockReturnValueOnce(false);
        fs.existsSync = mockExistsSync;
        let mockLstat = jest.fn();
        let isDirectory = jest.fn();
        isDirectory.mockReturnValue(true);
        mockLstat.mockReturnValue({
            isDirectory: isDirectory
        });
        fs.lstatSync = mockLstat;
        let mockGetFilesRecursively = jest.fn();
        mockGetFilesRecursively.mockReturnValueOnce([
            'file1.json'
        ]);
        tasksMod.getFilesRecursively = mockGetFilesRecursively;

        expect(() => {
            let tasks = tasksMod.traverseFiles(filePath);
        }).toThrowError('There was no file found at ' + filePath);
    });

    test("generateHTMLResults success with ignore exceptions", async () => {
        let mockHTML = {
            results: jest.fn().mockResolvedValue(mockPa11yResult)
        };
        tasksMod.html = mockHTML;
        Date.prototype.toString = jest.fn().mockReturnValue("Thu Jun 10 2021 18:46:30 GMT+0000 (Coordinated Universal Time)");

        let mockTask = {
            name: 'Test task',
            ignore: [
                "WCAG2AA.Principle1.Guideline1_1.1_1_1.H37",
                "WCAG2AA.Principle1.Guideline1_3.1_3_1.H49.AlignAttr",
                "WCAG2AA.Principle1.Guideline1_3.1_3_1.H43.IncorrectAttr",
                "WCAG2AA.Principle1.Guideline1_3.1_3_1.H43.MissingHeaderIds",
                "WCAG2AA.Principle1.Guideline1_3.1_3_1.H43.MissingHeadersAttrs",
                "WCAG2AA.Principle1.Guideline1_3.1_3_1.F92,ARIA4"
            ]
        };
        let html = await tasksMod.generateHTMLResults(mockPa11yResult, mockTask);
        expect(html).toMatchSnapshot();
    });

    test("generateHTMLResults success without ignore exceptions", async () => {
        let mockHTML = {
            results: jest.fn().mockResolvedValue(mockPa11yResult)
        };
        tasksMod.html = mockHTML;
        Date.prototype.toString = jest.fn().mockReturnValue("Thu Jun 10 2021 18:46:30 GMT+0000 (Coordinated Universal Time)");

        let mockTask = {
            name: 'Test task',
            ignore: []
        };
        let html = await tasksMod.generateHTMLResults(mockPa11yResult, mockTask);
        expect(html).toMatchSnapshot();
    });

    test("writeToFile success", async () => {
        fs.existsSync = jest.fn();
        fs.existsSync.mockReturnValue(false);
        let mockMakeDirSync = jest.fn();
        mockMakeDirSync.mockReturnValue(null);
        fs.mkdirSync = mockMakeDirSync;
        const spyMakeDirSync = jest
            .spyOn(fs, 'mkdirSync');
        let mockWriteFileSync = jest.fn();
        mockWriteFileSync.mockReturnValue(null);
        fs.writeFileSync = mockWriteFileSync;
        const spyWriteFileSync = jest
            .spyOn(fs, 'writeFileSync');
        const html = '<html>testData</html>';
        const directory = 'test/directory/path/';
        const filepath = directory + 'test.html';
        let htmlFilepath = await tasksMod.writeToFile(directory, filepath, html);
        expect(spyMakeDirSync).toHaveBeenCalledTimes(1);
        expect(spyWriteFileSync).toHaveBeenCalledTimes(1);
        expect(htmlFilepath).toContain('test/directory/path/test.html');
        spyMakeDirSync.mockRestore();
        spyWriteFileSync.mockRestore();
    });

    test("writeFileSync fail", async () => {
        const chalk = require ('chalk');
        fs.existsSync = jest.fn();
        fs.existsSync.mockReturnValue(true);
        let mockWriteFileSync = jest.fn();
        mockWriteFileSync.mockImplementation( () => {
            throw new Error('mockError');
        });
        fs.writeFileSync = mockWriteFileSync;
        const spyWriteFileSync = jest
            .spyOn(fs, 'writeFileSync');
        const spyConsoleError = spyOn(console, 'error');
        const html = '<html>testData</html>';
        const directory = 'test/directory/path/';
        const filepath = directory + 'test.html';
        let htmlFilepath = await tasksMod.writeToFile(directory, filepath, html);
        expect(htmlFilepath).toBe('mockError');
        expect(spyConsoleError).toHaveBeenCalledWith(chalk.red('mockError'));
        expect(spyWriteFileSync).toHaveBeenCalledTimes(1);
        spyWriteFileSync.mockRestore();
    });

    test("generateHTMLResults html.results fail", async () => {
        const chalk = require('chalk');
        const spyConsoleError = spyOn(console, 'error');
        await tasksMod.generateHTMLResults(null, 'task');
        expect(spyConsoleError).toHaveBeenCalledWith(chalk.red('Cannot read property \'documentTitle\' of null'));
    });

    test("msToTime", () => {
        let duration = 1800000;
        let timeStr = tasksMod.msToTime(duration);
        expect(timeStr).toBe('00:30:00.0');

        duration = 100;
        timeStr = tasksMod.msToTime(duration);
        expect(timeStr).toBe('00:00:00.1');

        duration = 3000;
        timeStr = tasksMod.msToTime(duration);
        expect(timeStr).toBe('00:00:03.0');

        duration = 30000;
        timeStr = tasksMod.msToTime(duration);
        expect(timeStr).toBe('00:00:30.0');

        duration = 300000;
        timeStr = tasksMod.msToTime(duration);
        expect(timeStr).toBe('00:05:00.0');

        duration = 3600000;
        timeStr = tasksMod.msToTime(duration);
        expect(timeStr).toBe('01:00:00.0');

        duration = 39600000;
        timeStr = tasksMod.msToTime(duration);
        expect(timeStr).toBe('11:00:00.0');
    });

    test("delay", async () => {
        let duration = 1000;
        let promise = tasksMod.delay(duration);
        setTimeout(() => {}, duration);
        expect(promise).resolves.toBe(expect.anything());

    })
});
;