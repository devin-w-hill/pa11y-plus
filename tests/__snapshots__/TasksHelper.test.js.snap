// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TasksHelper module tests generateHTMLResults success with ignore exceptions 1`] = `
"<!DOCTYPE html>
<html lang=\\"en\\">
<head>

	<meta charset=\\"utf-8\\"/>
	<title> Accessibility Report for Test task</title>

	<style>

		html, body {
			margin: 0;
			padding: 0;
			background-color: #fff;
			font-family: Arial, Helvetica, sans-serif;
			font-size: 16px;
			line-height: 22px;
			color: #333;
		}
		li {
			margin-bottom: 15px;
		}
		h2, h2, p, pre, ul {
			margin-top: 0;
			margin-bottom: 0;
		}
		h2 {
			margin-bottom: 10px;
			font-size: 24px;
			line-height: 24px;
		}
		h2 {
			font-size: 16px;
		}
		pre {
			white-space: pre-wrap;
			overflow: auto;
		}

		.page {
			max-width: 800px;
			margin: 0 auto;
			padding: 25px;
		}

		.counts {
			margin-top: 30px;
			font-size: 20px;
		}
		.count {
			display: inline-block;
			padding: 5px;
			border-radius: 5px;
			border: 1px solid #eee;
		}

		.clean-list {
			margin-left: 0;
			padding-left: 0;
			list-style: none;
		}
		.results-list {
			margin-top: 30px;
		}

		.result {
			padding: 10px;
			border-radius: 5px;
			border: 1px solid #eee;
		}
		.error {
			background-color: #fdd;
			border-color: #ff9696;
		}
		.warning {
			background-color: #ffd;
			border-color: #e7c12b;
		}
		.notice {
			background-color: #eef4ff;
			border-color: #b6d0ff;
		}

	</style>

</head>
<body>

	<div class=\\"page\\">

		<h1>Test task</h1>

		<h2>Accessibility Report For \\"https://govdemo.staging.tmorders.dev/portal/help/index\\"</h2>
		<p>Generated at: Thu Jun 10 2021 18:46:30 GMT+0000 (Coordinated Universal Time)</p>

		

		<div>
			<h4>Exceptions:</h4>
<ul>
		<li style=\\"margin-bottom:0px;\\">WCAG2AA.Principle1.Guideline1_1.1_1_1.H37</li>
		<li style=\\"margin-bottom:0px;\\">WCAG2AA.Principle1.Guideline1_3.1_3_1.H49.AlignAttr</li>
		<li style=\\"margin-bottom:0px;\\">WCAG2AA.Principle1.Guideline1_3.1_3_1.H43.IncorrectAttr</li>
		<li style=\\"margin-bottom:0px;\\">WCAG2AA.Principle1.Guideline1_3.1_3_1.H43.MissingHeaderIds</li>
		<li style=\\"margin-bottom:0px;\\">WCAG2AA.Principle1.Guideline1_3.1_3_1.H43.MissingHeadersAttrs</li>
		<li style=\\"margin-bottom:0px;\\">WCAG2AA.Principle1.Guideline1_3.1_3_1.F92,ARIA4</li>
</ul>
		</div>

		<p class=\\"counts\\">
			<span class=\\"count error\\">1 errors</span>
			<span class=\\"count warning\\">0 warnings</span>
			<span class=\\"count notice\\">0 notices</span>
		</p>

		<ul class=\\"clean-list results-list\\">
			<li class=\\"result error\\">
				<h2>Error: Iframe element requires a non-empty title attribute that identifies the frame.</h2>
				<p>WCAG2AA.Principle2.Guideline2_4.2_4_1.H64.1</p>
				<pre>&lt;iframe style=&quot;position: absolute; top: 0; left: 0; width: 100%; height: 100%;&quot; src=&quot;https://player.vimeo.com/video/398847860?color=ffffff&amp;amp;byline=0&amp;amp;portrait=0&quot; frameborder=&quot;0&quot; allowfullscreen=&quot;&quot;&gt;&lt;/iframe&gt; (select with \\"#orderModal &gt; div &gt; div &gt; div &gt; iframe\\")</pre>
			</li>
		</ul>

	</div>

</body>
</html>
"
`;

exports[`TasksHelper module tests generateHTMLResults success without ignore exceptions 1`] = `
"<!DOCTYPE html>
<html lang=\\"en\\">
<head>

	<meta charset=\\"utf-8\\"/>
	<title> Accessibility Report for Test task</title>

	<style>

		html, body {
			margin: 0;
			padding: 0;
			background-color: #fff;
			font-family: Arial, Helvetica, sans-serif;
			font-size: 16px;
			line-height: 22px;
			color: #333;
		}
		li {
			margin-bottom: 15px;
		}
		h2, h2, p, pre, ul {
			margin-top: 0;
			margin-bottom: 0;
		}
		h2 {
			margin-bottom: 10px;
			font-size: 24px;
			line-height: 24px;
		}
		h2 {
			font-size: 16px;
		}
		pre {
			white-space: pre-wrap;
			overflow: auto;
		}

		.page {
			max-width: 800px;
			margin: 0 auto;
			padding: 25px;
		}

		.counts {
			margin-top: 30px;
			font-size: 20px;
		}
		.count {
			display: inline-block;
			padding: 5px;
			border-radius: 5px;
			border: 1px solid #eee;
		}

		.clean-list {
			margin-left: 0;
			padding-left: 0;
			list-style: none;
		}
		.results-list {
			margin-top: 30px;
		}

		.result {
			padding: 10px;
			border-radius: 5px;
			border: 1px solid #eee;
		}
		.error {
			background-color: #fdd;
			border-color: #ff9696;
		}
		.warning {
			background-color: #ffd;
			border-color: #e7c12b;
		}
		.notice {
			background-color: #eef4ff;
			border-color: #b6d0ff;
		}

	</style>

</head>
<body>

	<div class=\\"page\\">

		<h1>Test task</h1>

		<h2>Accessibility Report For \\"https://govdemo.staging.tmorders.dev/portal/help/index\\"</h2>
		<p>Generated at: Thu Jun 10 2021 18:46:30 GMT+0000 (Coordinated Universal Time)</p>

		

		<div>
			<h4>Exceptions:</h4>
<ul>
		<li style=\\"margin-bottom:0px;\\">None</li>
</ul>
		</div>

		<p class=\\"counts\\">
			<span class=\\"count error\\">1 errors</span>
			<span class=\\"count warning\\">0 warnings</span>
			<span class=\\"count notice\\">0 notices</span>
		</p>

		<ul class=\\"clean-list results-list\\">
			<li class=\\"result error\\">
				<h2>Error: Iframe element requires a non-empty title attribute that identifies the frame.</h2>
				<p>WCAG2AA.Principle2.Guideline2_4.2_4_1.H64.1</p>
				<pre>&lt;iframe style=&quot;position: absolute; top: 0; left: 0; width: 100%; height: 100%;&quot; src=&quot;https://player.vimeo.com/video/398847860?color=ffffff&amp;amp;byline=0&amp;amp;portrait=0&quot; frameborder=&quot;0&quot; allowfullscreen=&quot;&quot;&gt;&lt;/iframe&gt; (select with \\"#orderModal &gt; div &gt; div &gt; div &gt; iframe\\")</pre>
			</li>
		</ul>

	</div>

</body>
</html>
"
`;
